### ISSUE TITLE
Limit to individual project directories

### ISSUE THREAD

**Issue by sschuberth** [tox-dev/pipdeptree#84]
It would be nice if the dependency reporting could be limited to a specific project directory. Usually, I'm not interested in all dependencies of all packages, but only in the (transitive) dependencies of a specific project, ideally even before having installed any of the dependencies.

**Comment by sschuberth** [comment]
To implement this, `pipdeptree` could internally probably do something like the following when being passed a project directory:

1. Create a virtualenv
2. Install pipdeptree in the virtualenv
3. Remember the installed packages in the virtualenv at this point
4. Install all packages required by the project in the virtualenv
5. Run `pipdeptree -l` in the virtualenv
6. Remove all packages from the list that were in 3. but are no project dependencies
7. Report the project dependencies

**Comment by naiquevin** [comment]
Not sure if I have understood what you mean by a `project directory`. But I am guessing you want to get all direct dependencies of a particular package, let's say `foo`. It can be achieved using the `-p` flag and `grep` as follows, 

```
$ pipdeptree -p foo | grep -E '^(foo|\s\s-)'
```
Here, `-p` will restrict the output to dependencies of `foo`. Then grep is used to match for two things:

1. Two spaces followed by a hyphen to match only those lines which are 1st level children (direct dependencies)
2. line containing the package name itself

This however is not possible without installing the package. The tmp virtualenv approach you've mentioned above should work though.

**Comment by sschuberth** [comment]
> But I am guessing you want to get all direct dependencies of a particular package

Not quite. By "project directory" I mean the root directory containing the source code of any Python project. Let's take https://github.com/nexB/scancode-toolkit as an example. After `git clone`, my Git working tree would be the project directory. Now I'd like to programmatically get the transitive dependency tree for this project. If you're familiar with Maven, that would be pretty much what `mvn dependency:tree` does.

**Comment by naiquevin** [comment]
If your repo has a `setup.py` file, you can try this: 

1. Create a new virtualenv
2. Run `pip install .` from the project root. 
3. Run the above command (pipdeptree + grep)

**Comment by sschuberth** [comment]
Right. I was just thinking that it would make sense to add this use case straight to `pipdeptree` so only a single command needs to be run. What do you think?

**Comment by naiquevin** [comment]
This is a very specific use case and I don't think the code for it belongs inside pipdeptree. But even if we decide to add it, it would be extremely tricky because pipdeptree works in a virtualenv only after it's manually installed inside it first. This makes the whole thing cyclic - pipdeptree creating virtualenv, then installing pipdeptree inside it and then replacing the current process with that executable.

Also, if you have the setup.py file, that itself would be the source of truth for first level dependencies of the project, right?

**Comment by sschuberth** [comment]
> [...] replacing the current process with that executable.

I don't think any such complex logic would be necessary. It would be just the global `pipdeptree` calling the one from the virtualenv, that's it.

> Also, if you have the setup.py file, that itself would be the source of truth for first level dependencies of the project, right?

Correct. But like I said, I'm interested in transitive dependencies as well. Also, I'd like to investigate arbitrary projects, so some might have a `requirements.txt` file instead of a `setup.py` file.

So feel free to close this if you believe this is out of scope.

**Comment by naiquevin** [comment]
> I don't think any such complex logic would be necessary. It would be just the global pipdeptree calling the one from the virtualenv, that's it.

This sounds like a better approach than replacing the current process.

> So feel free to close this if you believe this is out of scope.

Not sure if it's out of scope entirely, need to think about it. Will check `mvn dependency:tree`. Keeping the issue open.

**Comment by sschuberth** [comment]
Thanks for considering. FYI, see how the [docs for mvn dependency:tree](https://maven.apache.org/plugins/maven-dependency-plugin/tree-mojo.html) say "Displays the dependency tree for *this project*".

**Comment by gaborbernat** [comment]
I don't think we plan to support this for now.
